File: bindings/dart/test/client_test.dart
import 'dart:io';

import 'package:chrondb/chrondb.dart';
import 'package:test/test.dart';

bool _libAvailable = false;

void main() {
  late Directory tempDir;
  late ChronDB db;

  setUpAll(() {
    try {
      tempDir = Directory.systemTemp.createTempSync('chrondb-check-');
      final testDb = ChronDB.openPath(tempDir.path);
      testDb.close();
      tempDir.deleteSync(recursive: true);
      _libAvailable = true;
    } catch (_) {
      _libAvailable = false;
    }
  });

  setUp(() {
    if (!_libAvailable) return;
    tempDir = Directory.systemTemp.createTempSync('chrondb-test-');
    db = ChronDB.openPath(tempDir.path);
  });

  tearDown(() {
    if (!_libAvailable) return;
    db.close();
    tempDir.deleteSync(recursive: true);
  });

  test('put and get', () {
    if (!_libAvailable) return;
    db.put('user:1', {'name': 'Alice', 'age': 30});
    final result = db.get('user:1');
    expect(result['name'], equals('Alice'));
    expect(result['age'], equals(30));
  });

  test('get not found', () {
    if (!_libAvailable) return;
    expect(() => db.get('nonexistent:999'), throwsA(isA<DocumentNotFoundException>()));
  });

  test('delete', () {
    if (!_libAvailable) return;
    db.put('user:2', {'name': 'Bob'});
    db.delete('user:2');
    expect(() => db.get('user:2'), throwsA(isA<DocumentNotFoundException>()));
  });

  test('list by prefix', () {
    if (!_libAvailable) return;
    db.put('item:1', {'name': 'A'});
    db.put('item:2', {'name': 'B'});
    final result = db.listByPrefix('item');
    expect(result, isNotNull);
  });

  test('list by table', () {
    if (!_libAvailable) return;
    db.put('product:1', {'name': 'Widget'});
    final result = db.listByTable('product');
    expect(result, isNotNull);
  });

  test('history', () {
    if (!_libAvailable) return;
    db.put('doc:1', {'version': 1});
    db.put('doc:1', {'version': 2});
    final result = db.history('doc:1');
    if (result is List) {
      expect(result.length, greaterThanOrEqualTo(2));
    }
  });
}

File: bindings/go/client_test.go
package chrondb_test

import (
	"os"
	"path/filepath"
	"testing"

	chrondb "github.com/avelino/chrondb/bindings/go"
)

var libAvailable bool

func TestMain(m *testing.M) {
	dir, err := os.MkdirTemp("", "chrondb-check-")
	if err == nil {
		db, err := chrondb.OpenPath(dir)
		if err == nil {
			db.Close()
			libAvailable = true
		}
		os.RemoveAll(dir)
	}
	os.Exit(m.Run())
}

func setupDB(t *testing.T) (*chrondb.DB, string) {
	t.Helper()
	if !libAvailable {
		t.Skip("ChronDB shared library not available")
	}
	dir, err := os.MkdirTemp("", "chrondb-test-")
	if err != nil {
		t.Fatal(err)
	}
	db, err := chrondb.OpenPath(filepath.Join(dir, "db"))
	if err != nil {
		os.RemoveAll(dir)
		t.Fatal(err)
	}
	return db, dir
}

func TestPutAndGet(t *testing.T) {
	db, dir := setupDB(t)
	defer os.RemoveAll(dir)
	defer db.Close()

	doc := map[string]any{"name": "Alice", "age": float64(30)}
	_, err := db.Put("user:1", doc, "")
	if err != nil {
		t.Fatalf("put failed: %v", err)
	}

	result, err := db.Get("user:1", "")
	if err != nil {
		t.Fatalf("get failed: %v", err)
	}
	if result["name"] != "Alice" {
		t.Errorf("expected Alice, got %v", result["name"])
	}
}

func TestGetNotFound(t *testing.T) {
	db, dir := setupDB(t)
	defer os.RemoveAll(dir)
	defer db.Close()

	_, err := db.Get("nonexistent:999", "")
	if err != chrondb.ErrNotFound {
		t.Errorf("expected ErrNotFound, got %v", err)
	}
}

func TestDelete(t *testing.T) {
	db, dir := setupDB(t)
	defer os.RemoveAll(dir)
	defer db.Close()

	_, err := db.Put("user:2", map[string]any{"name": "Bob"}, "")
	if err != nil {
		t.Fatal(err)
	}

	err = db.Delete("user:2", "")
	if err != nil {
		t.Fatalf("delete failed: %v", err)
	}

	_, err = db.Get("user:2", "")
	if err != chrondb.ErrNotFound {
		t.Errorf("expected ErrNotFound after delete, got %v", err)
	}
}

func TestListByPrefix(t *testing.T) {
	db, dir := setupDB(t)
	defer os.RemoveAll(dir)
	defer db.Close()

	db.Put("item:1", map[string]any{"name": "A"}, "")
	db.Put("item:2", map[string]any{"name": "B"}, "")

	result, err := db.ListByPrefix("item", "")
	if err != nil {
		t.Fatalf("list by prefix failed: %v", err)
	}
	if result == nil {
		t.Error("expected non-nil result")
	}
}

func TestListByTable(t *testing.T) {
	db, dir := setupDB(t)
	defer os.RemoveAll(dir)
	defer db.Close()

	db.Put("product:1", map[string]any{"name": "Widget"}, "")

	result, err := db.ListByTable("product", "")
	if err != nil {
		t.Fatalf("list by table failed: %v", err)
	}
	if result == nil {
		t.Error("expected non-nil result")
	}
}

func TestHistory(t *testing.T) {
	db, dir := setupDB(t)
	defer os.RemoveAll(dir)
	defer db.Close()

	db.Put("doc:1", map[string]any{"version": float64(1)}, "")
	db.Put("doc:1", map[string]any{"version": float64(2)}, "")

	result, err := db.History("doc:1", "")
	if err != nil {
		t.Fatalf("history failed: %v", err)
	}
	if arr, ok := result.([]any); ok {
		if len(arr) < 2 {
			t.Errorf("expected at least 2 history entries, got %d", len(arr))
		}
	}
}
